extras/rebalance/directory-rebalance.py

```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/python3
+import argparse
+import os
+import errno
+import time
+import sys
+import datetime
+import hashlib
+import logging
+
+def size_fmt(num, suffix='B'):
+    for unit in ['','Ki','Mi','Gi','Ti','Pi','Ei','Zi']:
+        if abs(num) < 1024.0:
+            return "{:3.1f} {}{}".format(num, unit, suffix)
+        num /= 1024.0
+    return "{:.2f} {}{}".format(num, 'Yi', suffix)
+
+def time_fmt(fr_sec):
+    return str(datetime.timedelta(microseconds=fr_sec * 1000000))
+
+def crawl_progress(count, size):
+    sys.stdout.write('Building index of {} files with cumulative size {} to rebalance.\r'.format(count, size))
+    sys.stdout.flush()
+
+#https://gist.github.com/vladignatyev/06860ec2040cb497f0f3
+def progress(count, total, status=''):
+    bar_len = 60
+    filled_len = int(round(bar_len * count / float(total)))
+
+    percents = round(100.0 * count / float(total), 1)
+    bar = '=' * filled_len + '-' * (bar_len - filled_len)
+
+    sys.stdout.write('[{}] {}% ...{}\r'.format(bar, percents, status))
+    sys.stdout.flush()
+
+def progress_done():
+    print()
+
+class Rebalancer:
+    def __init__(self, path):
+        self.path = path # Path on which migration script is executed
+        self.migration_duration = 0 #Time spent in migrating data
+        self.skipped_migration_duration = 0 #Time spent in migrating data
+        self.migrated_files = 0 #Number of files migrated successfully
+        self.total_files = 0 #Total number of files scanned
+        self.total_size = 0 #Cumulative size of files scanned so far
+        self.expected_total_size = 0 #This is calculated at the time of populating index
+        self.expected_total_files = 0 #This is calculated at the time of populating index
+        self.migrated_size = 0 #Cumulative size of files migrated
+        self.index = self.get_file_name('index')
+        self.init_logging()
+
+    #Generate a unique name for the given path. format of the path will be
+    #rebalance-<hiphenated-path>-<first-8-hex-chars-of-md5-digest>.suffix
+    #If the length of this name is > 255 then hiphenated-path is truncated to
+    #make space
+    def get_file_name(self, suffix):
+        name_suffix = hashlib.md5(self.path.encode('utf-8')).hexdigest()[:8]+'.'+suffix
+        name_suffix = '-' + name_suffix
+        max_name_len = os.pathconf('.', 'PC_NAME_MAX')
+        name = 'rebalance'+self.path.replace('/', '-')
+        if len(name) > max_name_len - len(name_suffix):
+            name = name[:(max_name_len - len(name_suffix))]
+        name += name_suffix
```

Comment:
Is this really needed ? wouldn't a simple random unique file name be enough ?

Reply:
It would be to easy to associate the logfile for the particular directory we ran the script for. I wanted it to be same filename even if we run it multiple times for the same directory.